Checkboxes in "table-ripe" component

NFSS10 opened the pull request:
| - | - |
| --- | --- |
| Issue | - |
| Decisions | • Added checkboxes to the "table-ripe" component. Sorting the items doesn't affect the selection <br>• Ctrl + A = Selects all rows<br>• Ctrl + Alt + A = Deselects all rows <br>• Shift + Click and Shift + Click in another checkbox = Selects all the rows checkboxes between the two clicked rows |
| Animated GIF | ![table_checkboxes](https://user-images.githubusercontent.com/22588915/73004080-a35d2700-3dfe-11ea-91a0-d900aff94a0c.gif) <br><hr><br> ![table-ripe_checkboxes_shift_click](https://user-images.githubusercontent.com/22588915/74835155-a9b6c400-5314-11ea-8ebe-81bdceac3002.gif)|

BeeMargarida commented:
[BUG] When there is no items nor values, the global checkbox is checked

BeeMargarida commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -2,6 +2,14 @@
     <table class="table" v-bind:class="classes" v-bind:style="style">
         <thead class="table-head">
             <tr>
+                <th class="checkboxes-th" v-if="enableCheckboxes">
```
maybe "checkbox-global"

BeeMargarida commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -23,7 +31,18 @@
+                    v-bind:key="item.id"
+                    v-on:click="onRowClick(item, index)"
+                >
+                    <td class="checkbox-td" v-if="enableCheckboxes">
```
maybe "checkbox-item"

BeeMargarida commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -308,9 +377,73 @@ export const Table = {
+            if (itemsNrDiff >= 0) {
+                let item = null;
+                for (let i = items.length - itemsNrDiff; i < items.length; i++) {
+                    item = items[i];
```
maybe remove the "let item = null" from above and use "const item = item[i]" inside the loop

BeeMargarida commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -323,9 +415,55 @@ export const Table = {
+                this.globalCheckboxIcon = "check";
+                this.globalCheckboxValueData = true;
+            } else if (this.isAllUnchecked) {
+                this.$nextTick(() => {
```
I think it is not needed

NFSS10 commented on vue/components/ui/molecules/filter/filter.vue:
```diff
@@ -144,7 +159,8 @@ export const Filter = {
             loading: false,
-            tableTransition: ""
+            tableTransition: "",
+            checkedItemsData: this.checkedItems
```
Move closer to items

NFSS10 replied:
Init at the same time as items

NFSS10 commented on vue/components/ui/molecules/filter/filter.vue:
```diff
@@ -204,6 +220,19 @@ export const Filter = {
+                this.$emit("update:checked-items", value);
+            },
+            deep: true,
+            immediate: true
```
Same logic as `items`

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -1,7 +1,28 @@
-    <table class="table" v-bind:class="classes" v-bind:style="style">
+    <table
+        class="table"
+        v-bind:class="{ classes, disableSelection: lastClickedIndex !== null }"
```
move to classes

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -1,7 +1,28 @@
+        v-bind:style="style"
+    >
+        <global-events
+            v-on:keydown.meta.65.exact.prevent="onCtrlA()"
```
remove `()`

NFSS10 replied:
Make sure all prevents are needed

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -24,7 +45,28 @@
+                        v-bind:key="item.id"
+                        v-on:click.exact="onRowClick(item, index)"
+                        v-on:click.ctrl.exact="onRowCtrlClick(index, item.id)"
+                        v-on:click.meta.exact="onRowCtrlClick(index, item.id)"
```
onRowMetaClick

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -24,7 +45,28 @@
+                        <td class="checkbox-item" v-if="enableCheckboxes">
+                            <checkbox
+                                v-bind:size="8"
+                                v-bind:checked.sync="checkedItemsData[item.id]"
```
v-bind:checked="checkedItemsData[item.id]"
v-on:update:checked="value => checkedItemsData[item.id] = value"
v-on:update:checked="value => $set(checkedItemsData, item.id, value)" <-----

NFSS10 replied:
Since you're here, check ALL the `v-bind:SOMETHING.sync` for similar issues

NFSS10 replied:
@gcandal it is working because of the watch has deep as true (see below). We are also using watch this way in other components too, ex.: in "filter-ripe".

```
checkedItemsData: {
    immediate: true,
    deep: true,
    handler: function(value) {
        this.selectionChange();
        this.$emit("update:checked-items", value);
    }
}
```

NFSS10 replied:
I will use update either way, because when trying to uncheck all, `this.checkedItemsData = {};` the reactivity makes the the checkbox set the value to false.

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -115,6 +170,10 @@
     word-break: break-all;
 }
 
+.table .checkbox-item {
```
.table tbody > tr > td > checkbox-item,
.table table-head ....

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -277,39 +340,77 @@ export const Table = {
             type: Boolean,
             default: false
         },
+        transition: {
```
revert

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -277,39 +340,77 @@ export const Table = {
-            reverseData: this.reverse
+            reverseData: this.reverse,
+            globalCheckboxValueData: false,
+            globalCheckboxIcon: "check",
```
computed

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -323,9 +424,53 @@ export const Table = {
+        itemsWithIndex(items) {
+            return items.map((item, index) => ({ _originalIndex: index, ...item }));
+        },
+        initialCheckedItems() {
```
calculateCheckedItems

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -323,9 +424,53 @@ export const Table = {
+            const checkedItems = {};
+
+            this.items.forEach(item => {
+                checkedItems[item.id] = this.checkedItems[item.id]
```
checkedItems[item.id] |= false

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -336,8 +481,132 @@ export const Table = {
+            this.shiftIndex = null;
+            this.lastClickedIndex = null;
+        },
+        updateSelectionIndexes(index, itemId) {
```
try to find a more meaningful name and add docs if needed

NFSS10 commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -323,9 +424,53 @@ export const Table = {
+        initialCheckedItems() {
+            const checkedItems = {};
+
+            this.items.forEach(item => {
```
try to set `checkedItems` only when `true`

NFSS10 commented:
Removed:
• Ctrl + Click = Toogle clicked row checkbox
• Shift + Up = Starts selecting the rows checkboxes in the up direction
• Shift + Down = Starts selecting the rows checkboxes in the down direction

Changed "Shift + Click". it now only works in when doing it in the checkboxes and modified its behaviour to be equal to the one in gmail

joamag commented:
@gcandal maybe we can save this one if we use new naming (checkboxes is bad) and isolate/simplify the implementation

joamag commented:
Adding fast-track to close this one

joamag commented on vue/components/ui/organisms/listing/listing.stories.js:
```diff
@@ -1,5 +1,5 @@
 import { storiesOf } from "@storybook/vue";
-import { withKnobs, text, select } from "@storybook/addon-knobs";
+import { withKnobs, text, select, boolean } from "@storybook/addon-knobs";
```
Remove the checkbox demo

joamag commented on vue/components/ui/molecules/table/table.vue:
```diff
@@ -411,6 +488,48 @@ export const Table = {
+        onCheckboxClick(index, id) {
+            this.checkboxClick(index, id);
+        },
+        onCheckboxShiftClick(index, id) {
```
Remove this shift stuff
